### PR TITLE
layers: Make more PIPELINE_STATE fields be const

### PR DESCRIPTION
**Pull request by jeremyg-lunarg**
Restructure samplers_used_by_image and wrote_primitive_shading_rate be const.

### CODE REVIEW

File: layers/descriptor_sets.cpp [hunk 1]
```diff
@@ -1133,12 +1133,12 @@ bool CoreChecks::ValidateImageDescriptor(const char *caller, const DrawDispatchV
             static_cast<const cvdescriptorset::ImageSamplerDescriptor &>(image_descriptor).GetSamplerState());
     } else {
         if (binding_info.second.samplers_used_by_image.size() > index) {
-            for (auto &sampler : binding_info.second.samplers_used_by_image[index]) {
+            for (const auto &index : binding_info.second.samplers_used_by_image[index]) {
```
Comment (ziga-lunarg) [inline layers/descriptor_sets.cpp:1136]: Is this a good idea to use `index` and hide a function parameter?

Comment (jeremyg-lunarg) [reply layers/descriptor_sets.cpp:1136]: no! It is a really bad idea and I thought I fixed it. gcc 11 generates an error for this.

Comment (jeremyg-lunarg) [reply layers/descriptor_sets.cpp]: should be fixed now

**Review by ziga-lunarg — APPROVED** [review]

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -5746,20 +5746,6 @@ bool CoreChecks::PreCallValidateCreateGraphicsPipelines(VkDevice device, VkPipel
     return skip;
 }
 
-void CoreChecks::PostCallRecordCreateGraphicsPipelines(VkDevice device, VkPipelineCache pipelineCache, uint32_t count,
```
Comment (ncesario-lunarg) [inline layers/core_validation.cpp:5749]: 🎉

File: layers/descriptor_sets.cpp [hunk 3]
```diff
@@ -1133,12 +1133,12 @@ bool CoreChecks::ValidateImageDescriptor(const char *caller, const DrawDispatchV
...
         if (binding_info.second.samplers_used_by_image.size() > index) {
-            for (auto &sampler : binding_info.second.samplers_used_by_image[index]) {
+            for (const auto &desc_index : binding_info.second.samplers_used_by_image[index]) {
+                const auto *desc = descriptor_set->GetDescriptorFromBinding(desc_index.sampler_slot.binding, desc_index.sampler_index);
                 // NOTE: This check _shouldn't_ be necessary due to the checks made in IsSpecificDescriptorType in
                 //       shader_validation.cpp. However, without this check some traces still crash.
-                if (sampler.second && (sampler.second->GetClass() == cvdescriptorset::DescriptorClass::PlainSampler)) {
-                    const auto *sampler_state =
-                        static_cast<const cvdescriptorset::SamplerDescriptor *>(sampler.second)->GetSamplerState();
+                if (desc && (desc->GetClass() == cvdescriptorset::DescriptorClass::PlainSampler)) {
```
Comment (ncesario-lunarg) [inline layers/descriptor_sets.cpp:1140]: IIRC, it was never really understood why the `desc->GetClass() == cvdescriptorset::DescriptorClass::PlainSampler` check was necessary. On the one hand I wonder if this change allows this check to be removed. On the other hand, it's probably not worth the effort to find out as there's nothing incorrect about it.

**Review by ncesario-lunarg — APPROVED** [review]
LGTM!

File: layers/descriptor_sets.cpp (hunk 3 above)
Comment (jeremyg-lunarg) [reply layers/descriptor_sets.cpp:1140]: I don't get it either. I'm pretty sure I'll be messing with descriptors again (and again) soon, I'll try to remember this. The check is very cheap to run at least.

Comment (ncesario-lunarg) [reply layers/descriptor_sets.cpp:1140]: > The check is very cheap to run at least.

More curiosity as to what the original problem was, but yeah, it's not really costing anything, and especially not worth the effort if more refactorings are coming down the road.

**Pull request merged**